lib_eio_js/browser/eio_browser.ml

```diff
@@ -0,0 +1,198 @@
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+open Brr
+
+module Fiber_context = Eio.Private.Fiber_context
+module Run_queue : sig
+  type 'a t
+  (* A queue that supports pushing to the head of the queue *)
+
+  val create : unit -> 'a t
+
+  val push : 'a t -> 'a -> unit
+  (** [push q v] pushes a new item [v] to the back of the queue [q] *)
+
+  val push_head : 'a t -> 'a -> unit
+  (** [push_head q v] pushes a new item [v] to the head of the queue [q] *)
+
+  val pop : 'a t -> 'a option
+  (** [pop q] pops the next item of the queue if available. *)
+
+end = struct
+  type 'a t = 'a Lwt_dllist.t
+
+  let create () = Lwt_dllist.create ()
+
+  let push q v = ignore (Lwt_dllist.add_r v q : 'a Lwt_dllist.node)
+
+  let push_head q v = ignore (Lwt_dllist.add_l v q : 'a Lwt_dllist.node)
+
+  let pop q = Lwt_dllist.take_opt_l q
+end
+
+module Ctf = Eio.Private.Ctf
+
+module Suspended = struct
+  type 'a t = {
+    fiber : Eio.Private.Fiber_context.t;
+    k : ('a, unit) Effect.Deep.continuation;
+  }
+
+  let tid t = Eio.Private.Fiber_context.tid t.fiber
+
+  let continue t v =
+    Ctf.note_switch (tid t);
+    Effect.Deep.continue t.k v
+
+  let discontinue t ex =
+    Ctf.note_switch (tid t);
+    Effect.Deep.discontinue t.k ex
+end
+
+module Scheduler = struct
+  type t = {
+    listener : Brr.Ev.listener;
+    schedule : (unit -> unit);
+  }
+
+  let v ~schedule t =
+    let open Brr_io in
+    let scheduler = El.div [] in
+    let el = El.as_target scheduler in
+    let listener = Brr.Ev.listen Message.Ev.message (fun _ev -> schedule t) el in
+    let schedule =
+      let args = [| Ev.create Message.Ev.message |> Ev.to_jv |] in
+      let sched = El.to_jv scheduler in
+      fun () -> Jv.call sched "dispatchEvent" args |> ignore
+    in
+    { listener; schedule }
+
+  let stop t =
+    Brr.Ev.unlisten t.listener
+
+  let wakeup t = t.schedule ()
+end
+
+type t = {
+  (* Suspended fibers waiting to run again.
+     [Run_queue] is like [Stdlib.Queue], but allows pushing items
+     to the head too, which we need. *)
+  mutable run_q : (unit -> unit) Run_queue.t;
+  mutable pending_io : int;
+  mutable scheduler : Scheduler.t option;
+}
+
+let enqueue_thread t k v =
+  Run_queue.push t.run_q (fun () -> Suspended.continue k v);
+  t.pending_io <- t.pending_io - 1;
+  Option.iter Scheduler.wakeup t.scheduler
+
+let enqueue_failed_thread t k v =
+  Run_queue.push t.run_q (fun () -> Suspended.discontinue k v);
+  t.pending_io <- t.pending_io - 1;
+  Option.iter Scheduler.wakeup t.scheduler
+
+
+type _ Effect.t += Enter_unchecked : (t -> 'a Suspended.t -> unit) -> 'a Effect.t
+let enter_unchecked fn = Effect.perform (Enter_unchecked fn)
+
+let enter_io fn =
+  enter_unchecked @@ fun st k ->
+  st.pending_io <- st.pending_io + 1;
+  fn st k
+
+(* Resume the next runnable fiber, if any. *)
+let schedule t : unit =
+  match Run_queue.pop t.run_q with
+  | Some f -> f ();
```

Comment:
Why doesn't this need to call `schedule` again after `f` is done?

Reply:
Yes, good question! 

So after a bit of head scratching, I think the answer is it should call schedule again but because of how everything is implemented it doesn't actually need too... maybe. So I think this is because the only time we could potentially hang is if there is more than one item in the run_q *and* there is no scheduled wakeup call. I think the only way to get more than one thing in the queue is if you do some "IO" like the timeout (so not by forking, awaiting promises etc.) and because every single "IO" operation sends a wakeup event to our scheduler then the run_q will always be fully processed (but perhaps more slowly). At any rate, I've added a call to schedule here!

Reply:
We need to be careful here. In `Eio_linux`, `schedule` returns ``[`Exit_scheduler]`` and a suspended fiber has type ``('a, [`Exit_scheduler]) continuation``. This indicates that resuming a continuation commits to keeping things going (i.e. calling `schedule` again).

Your `run_q` has type `(unit -> unit) Run_queue.t`, which suggests it doesn't do that.

It would probably be best to decide which one you're doing and update the types to check it. If you call `schedule` when you don't need to, you can easily make your continuation handler not be tail-recursive, which is bad.

Reply:
Good point! So in https://github.com/ocaml-multicore/eio/pull/405/commits/9ff3108b7156a799f228ea85e505bb3b9a145e73 I've changed the scheduler again (sorry...). It is now much more reminiscent of the luv scheduler modulo `Luv.Async.send` is now a call to `requestIdleCallback` the idea being anytime we fall through to the final "exit" promise we'll be idle and allow the scheduler to be woken up. This means `schedule` (now `wakeup`) is only ever called when needed. Unfortunately safari doesn't implement `requestIdleCallback` but this seems like a shim people have used in their libraries. I do note that react.js used to use it in their scheduler but I think decided that it wasn't aggressive enough, I don't know if that's a problem or not here.

Reply:
I think this may suffer from the same problem as #427